hadoop-ozone/s3gateway/src/main/java/org/apache/hadoop/ozone/s3/exception/S3ErrorTable.java

```diff
@@ -122,6 +122,9 @@ private S3ErrorTable() {
       "NotImplemented", "This part of feature is not implemented yet.",
       HTTP_NOT_IMPLEMENTED);
 
+  public static final OS3Exception NO_OVERWRITE = new OS3Exception(
+      "Conflict", "Cannot overwrite file with directory", HTTP_CONFLICT);
```

Comment:
It looks like there is no failure when writing a directory when one already exists. If it is just a no-op then this message is accurate because the no-overwrite case will only happen when a file is already there. However it may be good to manually verify that doing directory create from s3g with this PR's changes when a directory already exists there doesn't result in 500 or other obscure error.

Reply:
I've tested it manually already. This is no-op. Here are the results:
```
bash-4.2$ ozone sh bucket create /s3v/fso-bucket -l FILE_SYSTEM_OPTIMIZED

ozone git:(HDDS-7594) ✗ aws s3api put-object --endpoint http://localhost:9878 --bucket fso-bucket --key dir1

bash-4.2$ ozone sh key list /s3v/fso-bucket
[ {
  "volumeName" : "s3v",
  "bucketName" : "fso-bucket",
  "name" : "dir1/",
  "dataSize" : 0,
  "creationTime" : "2023-02-06T12:01:51.883Z",
  "modificationTime" : "2023-02-06T12:01:51.883Z",
  "replicationConfig" : {
    "replicationFactor" : "ONE",
    "requiredNodes" : 1,
    "replicationType" : "RATIS"
  },
  "metadata" : { }
} ]

ozone git:(HDDS-7594) ✗ aws s3api put-object --endpoint http://localhost:9878 --bucket fso-bucket --key dir1

bash-4.2$ ozone sh key list /s3v/fso-bucket
[ {
  "volumeName" : "s3v",
  "bucketName" : "fso-bucket",
  "name" : "dir1/",
  "dataSize" : 0,
  "creationTime" : "2023-02-06T12:01:51.883Z",
  "modificationTime" : "2023-02-06T12:01:51.883Z",
  "replicationConfig" : {
    "replicationFactor" : "ONE",
    "requiredNodes" : 1,
    "replicationType" : "RATIS"
  },
  "metadata" : { }
} ]

ozone git:(HDDS-7594) ✗ aws s3api put-object --endpoint http://localhost:9878 --bucket fso-bucket --key dir1/key1 --body README.md 

bash-4.2$ ozone sh key list /s3v/fso-bucket
[ {
  "volumeName" : "s3v",
  "bucketName" : "fso-bucket",
  "name" : "dir1/",
  "dataSize" : 0,
  "creationTime" : "2023-02-06T12:01:51.883Z",
  "modificationTime" : "2023-02-06T12:01:51.883Z",
  "replicationConfig" : {
    "replicationFactor" : "ONE",
    "requiredNodes" : 1,
    "replicationType" : "RATIS"
  },
  "metadata" : { }
}, {
  "volumeName" : "s3v",
  "bucketName" : "fso-bucket",
  "name" : "dir1/key1",
  "dataSize" : 4263,
  "creationTime" : "2023-02-06T12:03:08.380Z",
  "modificationTime" : "2023-02-06T12:03:09.141Z",
  "replicationConfig" : {
    "replicationFactor" : "ONE",
    "requiredNodes" : 1,
    "replicationType" : "RATIS"
  },
  "metadata" : { }
} ]

ozone git:(HDDS-7594) ✗ aws s3api put-object --endpoint http://localhost:9878 --bucket fso-bucket --key dir1

bash-4.2$ ozone sh key list /s3v/fso-bucket
[ {
  "volumeName" : "s3v",
  "bucketName" : "fso-bucket",
  "name" : "dir1/",
  "dataSize" : 0,
  "creationTime" : "2023-02-06T12:01:51.883Z",
  "modificationTime" : "2023-02-06T12:01:51.883Z",
  "replicationConfig" : {
    "replicationFactor" : "ONE",
    "requiredNodes" : 1,
    "replicationType" : "RATIS"
  },
  "metadata" : { }
}, {
  "volumeName" : "s3v",
  "bucketName" : "fso-bucket",
  "name" : "dir1/key1",
  "dataSize" : 4263,
  "creationTime" : "2023-02-06T12:03:08.380Z",
  "modificationTime" : "2023-02-06T12:03:09.141Z",
  "replicationConfig" : {
    "replicationFactor" : "ONE",
    "requiredNodes" : 1,
    "replicationType" : "RATIS"
  },
  "metadata" : { }
} ]

ozone git:(HDDS-7594) ✗ aws s3api put-object --endpoint http://localhost:9878 --bucket fso-bucket --key dir1/key1                 

An error occurred (Conflict) when calling the PutObject operation: Cannot overwrite file with directory
```

According to those above, everything is working as intended.

Reply:
Thanks for checking, that looks good.